apps/term-apply/pkg/applicant/manager.go

```diff
@@ -66,91 +60,91 @@ func (a *ApplicantManager) AddApplicant(userID, name, email string, role int) er
 			}
-			return nil
 		}
+		return nil
+	} else {
+		log.Printf("Creating new application for applicant %s with (%s, %s, %s)", github, name, email, roleStr)
+		a.mu.Unlock()
+		a.writeChan <- applicationPacket{app: newApplication, updateInPlace: false}
+		return nil
 	}
-
-	log.Printf("Adding new applicant %s with (%s, %s, %s)", userID, name, email, roleStr)
-	a.applicants = append(a.applicants, newApplicant)
-	a.mu.Unlock()
-	a.writeChan <- a.applicants
-	return nil
 }
 
-func (a *ApplicantManager) readDataFile(filename string) error {
-	s3file.CopyFromS3(a.bucket, a.csvKey, filename)
-	records, err := readData(filename)
-	if err != nil {
-		return err
-	}
-
-	a.mu.Lock()
-	defer a.mu.Unlock()
-	a.applicants = []applicant{}
-	for _, record := range records {
-		if len(record) != 4 {
-			return fmt.Errorf("invalid record %v, fix csv file", record)
-		}
-		a.applicants = append(a.applicants,
-			applicant{
-				userID: record[0],
-				name:   record[1],
-				email:  record[2],
-				role:   record[3],
-			})
-	}
-	return nil
-}
-
-func (a *ApplicantManager) writeDataFile(filename string, writeChan chan []applicant) error {
+func (a *ApplicantManager) writeDynamoItem(writeChan chan applicationPacket) {
 	for {
-		updatedApplicants := <-writeChan
-		a.mu.RLock()
-		records := [][]string{}
-		for _, applicant := range updatedApplicants {
-			records = append(records, []string{
-				applicant.userID,
-				applicant.name,
-				applicant.email,
-				applicant.role,
-			})
-		}
-		a.mu.RUnlock()
-
-		if err := writeData(filename, records); err != nil {
-			log.Printf("error writing file %s, %v", filename, err)
-			return err
-		}
-
-		if err := s3file.CopyToS3(a.bucket, filename, a.csvKey); err != nil {
-			log.Printf("error writing to s3 %s, %s, %v", filename, a.csvKey, err)
+		packet := <-writeChan
+
+		if packet.updateInPlace {
+			if packet.prevEmail == packet.app.email {
+				log.Printf("Updating dynamodb record in %s for %s", a.dynamodbTable, packet.app.github)
+				if err := UpdateApplication(packet.app, a.dynamodbTable); err != nil {
+					log.Printf("Error uploading application for %s in %s: %v", packet.app.github, a.dynamodbTable, err)
+				} else {
+					log.Printf("Succesful write")
+				}
+			} else {
+				log.Printf("Deleting and recreating dynamodb record in %s for %s", a.dynamodbTable, packet.app.github)
+				if err := RecreateApplication(packet.app, packet.prevEmail, a.dynamodbTable); err != nil {
+					log.Printf("Error uploading application for %s in %s: %v", packet.app.github, a.dynamodbTable, err)
+				} else {
+					log.Printf("Succesful write")
+				}
+			}
+		} else {
+			log.Printf("Writing new record to dynamodb in %s for %s", a.dynamodbTable, packet.app.github)
+			if err := PutApplication(packet.app, a.dynamodbTable); err != nil {
+				log.Printf("Error uploading application for %s in %s: %v", packet.app.github, a.dynamodbTable, err)
+			} else {
+				log.Printf("Succesful write")
+			}
 		}
 	}
 }
```

Comment:
suggest something like this for better maintainability and less cognitive load for future contributors

```
for {
	packet := <-writeChan

	// new application
	if !packet.updateInPlace {
		log.Printf("Writing new record to dynamodb in %s for %s", a.dynamodbTable, packet.app.github)
		if err := PutApplication(packet.app, a.dynamodbTable); err != nil {
			log.Printf("Error uploading application for %s in %s: %v", packet.app.github, a.dynamodbTable, err)
			continue
		}
		log.Printf("Succesful write")
		continue
	}

	// update existing application with same email address
	if packet.prevEmail == packet.app.email {
		log.Printf("Updating dynamodb record in %s for %s", a.dynamodbTable, packet.app.github)
		if err := UpdateApplication(packet.app, a.dynamodbTable); err != nil {
			log.Printf("Error uploading application for %s in %s: %v", packet.app.github, a.dynamodbTable, err)
			continue
		}
		log.Printf("Succesful write")
		continue
	}

	// delete and recreate application due to update email address (email is primary key)
	log.Printf("Deleting and recreating dynamodb record in %s for %s", a.dynamodbTable, packet.app.github)
	if err := RecreateApplication(packet.app, packet.prevEmail, a.dynamodbTable); err != nil {
		log.Printf("Error uploading application for %s in %s: %v", packet.app.github, a.dynamodbTable, err)
		continue
	}
	log.Printf("Succesful write")
	}
}
```

Reply:
This I'm definitely going to use a switch statement for. I'm going to change the `UpdateInPlace` variable to be an enum type that corresponds to what kind of write needs to happen. This would be much simpler in the long run, as the function that writes to the channel can more easily handle that logic.

Reply:
Yeah, the switch statement makes things look MUCH nicer. I don't think there will be more types of writing necessary, but in case there are, it would be easy enough to add to the enum and switch statement